polyply/src/gmx_energy_minize.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2020 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import subprocess
+from pathlib import Path
+import tempfile
+import vermouth
+from .dispatch_gmx import DispatchGMX
+
+
+def _find_max_force(mdrun_out):
+    for line in mdrun_out.split("\n"):
+        if "Maximum force" in line:
+            force = line.split()[3]
+    return float(force)
+
+class GMXMinimize():
+    """
+    Processor for energy minimizing a system
+    """
+
+    def __init__(self, top_file, mdp_file, max_force=10**3.0, max_cycles=10):
+        self.top_file = Path(top_file).absolute()
+        self.mdp_file = Path(mdp_file).absolute()
+        self.max_cycles = max_cycles
+        self.max_force = max_force
+        self.top_dir = os.getcwd()
+
+    def run_topology(self, topology):
+
+        tmpdir = tempfile.TemporaryDirectory(dir=self.top_dir)
+        mdrun = DispatchGMX("mdrun", tmpdir.name, "gmx")
+        grompp = DispatchGMX("grompp", tmpdir.name, "gmx")
```

Comment:
It looks here like DispatchGMX could be a factory function/class instead. Ideally one that creates Singletons, so there's only ever one (1) MDRun class. The latter does shift the cwdir responsibility to the caller, but that can be caught by e.g.
```
mdrun = MDRun()
mdrun.run(cwdir='/my/path/', s=tpr_file, nsteps=100)
```
Note that the MDRun (and other GMX classes) take care of turning the arguments to str.